### PR TITLE
Improvements regarding signed URLs

### PR DESCRIPTION
**Pull request by m-mohr**
Solves issue #379

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -2980,26 +2980,27 @@ paths:
...
         programs, e.g. web browsers or download managers. Back-ends are
         responsible to generate the URL signing keys and to manage their
         appropriate expiration. The back-end MAY indicate an expiration time by
-        setting the `expires` property.
-
+        setting the `expires` property in the reponse. Requesting this endpoint
+        SHOULD always return valid URLs. If the links generated for a previous
+        request have expired, the signed URLs SHOULD be renewed. 
+        Once the signed URLs have expired, the signed URLs MAY return the openEO
+        error `LinkExpired`.
```
Comment (soxofaan) [inline openapi.yaml:3003]: It took me some re-reading to fully understand the points you're making here.

fine-tuning suggestion:

> Requesting this endpoint SHOULD always return URLs that are valid within a reasonable time window.
> Signed URLs that were generated for a previous request and already expired SHOULD NOT be reused, but regenerated with new expiration time.
> Signed URLs that expired MAY return the openEO error `LinkExpired`.

**Comment by soxofaan** [comment]
After pondering some more on this, some thoughts:

- shouldn't we be more enforcing about the expiration time for security reasons (e.g. at least explicitly RECOMMEND it)? The current statement "Back-ends are responsible  ..." is weaker than a recommendation I think.
- Would it make sense to let the user choose whether they want public signed URLs that can expire or URLs protected with bearer token auth (that can not expire)? The expiry problem is mitigated by the renewal mechanism, but security minded users might want to avoid  publicly available URLs that they can not invalidate themself.

I know, the tin foil hat factor of these points is quite high, and it's probably not urgent to tackle these the moment, but it might become a problem in the future.

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `GET /credentials/oidc` can provide a set of default client ids for OpenID Connect. [#366](https://github.com/Open-EO/openeo-api/pull/366)
 - `experimental` and `deprecated` flags added for file formats, service types, udf runtimes, udf runtime versions, udf runtime libraries and all related parameters and schemas. [#354](https://github.com/Open-EO/openeo-api/issues/354)
 - `GET /jobs/{job_id}/logs` and `GET /services/{service_id}/logs`: `time` property added. [#330](https://github.com/Open-EO/openeo-api/issues/330)
+- Added error `FileExpired`.  [#379](https://github.com/Open-EO/openeo-api/issues/379)
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply CHANGELOG.md:14]: Will fix.

File: errors.json [hunk 3]
```diff
@@ -290,6 +290,14 @@
 			"Batch Jobs"
 		]
 	},
+	"LinkExpired": {
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply errors.json:293]: Changed to "ResultLinkExpired".

File: errors.json [hunk 4]
```diff
@@ -290,6 +290,14 @@
...
+		"description": "The signed URLs for batch job results have expired. Please send a request to `GET /jobs/{job_id}/results` to refresh the links.",
+		"message": "The links to the batch job results have expired. Please request the results again.",
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply errors.json:295]: Changed.

File: openapi.yaml (hunk 1 above)
Comment (m-mohr) [reply openapi.yaml:3003]: Mostly adopted your proposal.

File: openapi.yaml [hunk 5]
```diff
@@ -2980,26 +2980,27 @@ paths:
...
         URL signing is a way to protect files from unauthorized access with a
         key in the URL instead of HTTP header based authorization. The URL
         signing key is similar to a password and its inclusion in the URL allows
         to download files using simple GET requests supported by a wide range of
         programs, e.g. web browsers or download managers. Back-ends are
         responsible to generate the URL signing keys and to manage their
         appropriate expiration. The back-end MAY indicate an expiration time by
-        setting the `expires` property.
-
+        setting the `expires` property in the reponse. Requesting this endpoint
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply openapi.yaml:2999]: Hmm... I'm not sure how high the security risk is actually. Google uses signed URLs for Docs, Spreadsheets etc. and they seem to not expire at all.

**Comment by m-mohr** [comment]
> * shouldn't we be more enforcing about the expiration time for security reasons (e.g. at least explicitly RECOMMEND it)? The current statement "Back-ends are responsible  ..." is weaker than a recommendation I think.

As also stated above, I'm not sure how high the security risk is actually, so I'm not sure this is actually required.

> * Would it make sense to let the user choose whether they want public signed URLs that can expire or URLs protected with bearer token auth (that can not expire)?

I assume that could make sense, but that would need to be tracked in a new issue for 2.0 (breaking for clients).

> The expiry problem is mitigated by the renewal mechanism

The wording we add here doesn't necessarily mean that previous signed URLs get invalid, it just means expired links will be renewed. Revoking signed URLs is a different issue: #341 We may decide that re-requesting this endpoint also invalidates previously generated signed URLs, but currently that's not the intention yet.

File: openapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (soxofaan) [reply openapi.yaml:2999]: a difference is that the google doc user can stop sharing a doc (invalidate the signed url)

[the remark this answers is not in the document]
Comment (m-mohr) [reply openapi.yaml:2999]: Yes, that is what issue #341 has been opened for.

**Comment by soxofaan** [comment]
I didn't know about #341 , that should indeed tackle most of the concerns raised here.

**Comment by m-mohr** [comment]
@soxofaan It seems this is important for you? I can certainly move the milestone up for #341 and make a proposal for API 1.1 in the next days.

**Comment by soxofaan** [comment]
Well, to give some background: the current implementation that @laxiwuka is working on for the VITO backend does not allow invalidation (there is no storage, just an added hash in the URL based on a secret to check that the URL hasn't been tampered with). 
The only tool to improve security is using shorter expiration times (changing the secret would invalidate all URLs for all users).

#341  requires more work (because of persistent storage involved) than other alternatives (that do not require persistent storage or state) I am exploring here:
- push the spec to make  expiration times as short as possible
- make it possible to opt-in for bearer-token based urls
- allow user or client to set expiration time in some way

**Comment by m-mohr** [comment]
Thanks for the details, @soxofaan. It sounds like this is not your long-term solution.

I'll merge this and have opened PR #381 with a proposal to revoke signed URLs.

**Comment by soxofaan** [comment]
Yes, good to merge this already. 
Another follow up about toggling between signed urls and  bearer token download urls: #382

**Pull request merged**